include/ProgramRepresentation/PVAliasSet.h

```diff
@@ -0,0 +1,71 @@
+#ifndef PV_ALIAS_SET_H
+#define PV_ALIAS_SET_H
+
+#include "ProgramRepresentation/MethodsSet.h"
+#include "ProgramRepresentation/ProgramVariable.h"
+#include "Utils.h"
+
+// holds a set of program variables and methods called on them.
+// each program variable in the set is considered an alias set.
+class PVAliasSet {
+private:
+  std::list<ProgramVariable> programVariables;
+
+  // holds the methods called on the set
+  MethodsSet methods;
+
+public:
+  PVAliasSet();
+
+  // returns true iff programVar is in the set by checking its raw name
+  bool contains(ProgramVariable programVar);
+
+  // returns true iff there exists a program variable in our set with raw name
+  // equal to rawName
+  bool containsByRawName(const std::string &rawName);
+
+  // returns true iff there exists a program variable in our set with clean name
+  // equal to cleanedName
+  bool containsByCleanedName(const std::string &cleanedName);
+
+  // adds programVar to our set of program variables. programVar is checked to
+  // see if it already exists in our set of program variables and it's not added
+  // if it already does.
+  void add(ProgramVariable programVar);
+
+  // adds array of program variables to our set of program variables.
+  // each variable is checked to see if already exists in our list of program
+  // variables.
+  void addProgramVariables(std::list<ProgramVariable> programVariables);
+
+  std::list<ProgramVariable> getProgramVariables();
+
+  MethodsSet getMethodsSet() const;
+  MethodsSet *getMethodsSetRef();
+
+  void setMethodsSet(MethodsSet methods);
```

Comment:
I'm not sure if you're referring to both getMethodsSetRef and setMethodsSet, but I'll explain my use for both. 

`getMethodsSetRef` is used a number of times in DataflowPass.cpp when we insert an alias set's methods into a call like `onAllocationFunctionCall`, `onReallocFunctionCall`, `onSafeFunctionCall`, `onAnnotation`, etc.. They take `MethodsSet*` as input and either call `addMethod` or `clearMethods`. Direct reference to the methods set itself is needed so I can modify the methods set accordingly; if I were to use `getMethodsSet` instead, I'd just get a copy of the methods set. 

There are two places where `setMethodsSet` is useful: 

1) [lines 263 – 285 of DataflowPass](https://github.com/Kevin-TD/c-resource-leak-checker/blob/f5b3950358b9f0a4d29a04e9fa51b9fd67f8a6de/src/DataflowPass.cpp#L263C13-L263C13)

Here, we're computing a least upper bound, and it's easier to just take the existing `pvas` and change its methods set, rather than clone `pvas` and insert that `lubMethodsSet`. 

2) [lines 138 – 143 of TestRunner ](https://github.com/Kevin-TD/c-resource-leak-checker/blob/f5b3950358b9f0a4d29a04e9fa51b9fd67f8a6de/src/TestRunner.cpp#L138) where we build the expected result 

Every getter is building part of the result as the 2nd parameter (`addNewIfNotFound`) is set to true. We can then easily insert the methods set.

Reply:
I see your point about the need to modify the set - I was not disputing that it's useful. However, I do think exposing this raw pointer as part of the API is asking for trouble. In particular, I think the LUB logic should be inside of `PVAliasSet` rather than in `DataflowPass`: `DataflowPass` should call some function named `lub` or similar that does the same computation, but internal to the `PVAliasSet` class. That way, arbitrary clients later on (i.e., us) won't be able to make **arbitrary** changes to the set: it'll only be possible to lub.

For the `TestRunner` case, it's less obvious. I would handle that with `friend` rather than exposing the underlying data structure to all clients.